Implement UNBUFFERED_PRINT option

eclecticdave opened the pull request:
As discussed, this PR is the first part of PR #6003 split out and includes the implementation of rawPrint and rawPrintErr.

These methods print to stdout and stderr without requiring strings to be terminated with newlines. They have been added to the Module object for Node and to shell[_minimal].html for browsers. Unfortunately spidermonkey has no equivalent of putstr that prints to stderr so only rawPrint is implemented there.

In addition tty.read has been modified slightly to return when a whole line has been entered, rather than waiting for EOF.  On Node and SpiderMonkey, this makes input from stdin (e.g. scanf) work the same way as native terminals when run interactively.

I've added a new test to the 'others' suite - test_stdin2.  This is a simple prompt and response program which demonstrates the changes.  Although it is setup to run under the test suite, it's really better run interactively, since it isn't so much important what it prints as when it flushes the output to the screen and running it as an automated test doesn't really capture that subtlety.  It runs best interactively under Node as the lack of rawPrintErr under Spidermonkey causes the last prompt to not be flushed until after the response is entered.

* Implement rawPrint and rawPrintErr (see Issue #4366)
* Some tests need modifying due to output changing validly because of
  rawPrint:
  - test_exceptions_uncaught_2 and test_remove - std:cout now no longer
    automatically appends an eol, so added '<< std:endl' as needed.
  - test_fs_trackingdelegate - printf, for reasons known only to itself,
    splits its calls to 'write' into two, the first 26 characters followed
    by the remaining 2 chars.  Since rawPrint does not buffer data across
    calls to 'write', this results in the text 'Wrote to file "/dev/tty"'
    getting output after the first 26 characters have been output.
  - test_printf - line 'no_new_line' is now not followed with a new line,
    this is valid - modified output.txt accordingly.
  - test_rounding, test_simd15 - last printf now correctly doesn't emit a
    new line - .out files modified accordingly.
* Fix to make sure tty read returns on EOL, instead of requiring EOF.
* New test test_stdin2 added.

eclecticdave commented:
As discussed in PR #6082 - I've replaced rawPrint and rawPrintErr with 'printChars'.  The browser version  is basically @juj's 'writeOut' function with a tweak for UTF8.  As mentioned in the comments on that PR, I've decided I like the idea of a single function with an FD parameter as it leaves the door open for multiple output streams to be easily implemented later.

juj commented on src/library_tty.js:
```diff
@@ -63,13 +63,24 @@ mergeInto(LibraryManager.library, {
       },
       write: function(stream, buffer, offset, length, pos) {
+        if (Module['printChars'] != undefined) {
+          Module['printChars'](stream.fd, buffer.subarray(offset, offset + length), length);
```
Instead of `buffer.subarray(offset, offset + length)`, which allocates garbage, I think it would be better to define the function to go for something like `Module['printChars'](stream.fd, buffer, offset, length);`?

eclecticdave replied:
This is because as it stands UTF8ArrayToString is expecting to be passed a nul terminated string.  A subarray works as well because an out-of-bounds index ref returns undefined.  It might be more efficient to have an optional 'length' parameter on UTF8ArrayToString, maybe?

Unless we can assume we'll only be calling printChars with nul terminated strings in the first place, but then having a 'length' parameter implies that's not necessarily the case?

eclecticdave replied:
Latest comment adds optional 'len' parameter to UTF8ArrayToString

juj commented on src/library_tty.js:
```diff
@@ -63,13 +63,24 @@ mergeInto(LibraryManager.library, {
           buffer[offset+i] = result;
+
+          // If EOL character, return having put it in the buffer.
+          if (result === 10) break;
```
Hmm, why this?

eclecticdave replied:
Currently 'read' doesn't return until you hit CTRL-D to send EOF (or fill the buffer).  This is not obvious from the automated tests, but clearly isn't what you want from an interactive session, where it should return when you press RETURN.

juj replied:
Thanks, this I think is something we should have an interactive test for, it is quite a specific scenario.

juj commented on src/library_tty.js:
```diff
@@ -80,6 +91,9 @@ mergeInto(LibraryManager.library, {
         }
+        // Sending null causes output to be printed, even if it didn't end in a
+        // newline.
+        stream.tty.ops.put_char(stream.tty, null);
```
I'm not sure I follow why this was added? Was there something that did not work before?

eclecticdave replied:
Oops, I'd meant to be remove that!  It's left over from the previous commit, when I was sending each character through 'put_char' and signalling I'd finished by sending a null.  Since I've changed it to call printChars further up, this is no longer needed.  I'll take it out.

juj commented on src/library_tty.js:
```diff
@@ -131,7 +145,8 @@ mergeInto(LibraryManager.library, {
             // Browser.
-            result = window.prompt('Input: ');  // returns null on cancel
+            var result;
+            result = window.prompt('Input: ');   // returns null on cancel
```
This could be just `var result = window.prompt('Input: ');`?

eclecticdave replied:
Yeah, in fact there's already a 'var result  = null' at the top of the function, so this line isn't needed at all.

juj commented on src/library_tty.js:
```diff
@@ -168,9 +183,9 @@ mergeInto(LibraryManager.library, {
-          tty.output = [];
-        } else {
-          if (val != 0) tty.output.push(val);
+          tty.output = [];         
```
(needless trailing whitespace)

eclecticdave replied:
Yep, will remove.

juj commented on src/preamble.js:
```diff
@@ -433,7 +433,15 @@ function UTF8ArrayToString(u8Array, idx) {
+    var buf;
+#ifdef USE_PTHREADS
+    // Need to copy the data to a non-shared buffer as TextDecoder isn't compatible with SharedArrayBuffer.
+    // Uint8Array.from would be more readable, but isn't universally available.
```
Hmm, I think this is a Chrome bug? There was a doc somewhere about which APIs should be SAB compatible, and I think I did read TextDecoder there, and iirc I did test that to work on Firefox. If we had a `MINIMUM_CHROME_VERSION` #ifdef we've been talking about, that could be used here, although right now probably nothing to change on this.

eclecticdave replied:
I was seeing a specific error message along the lines of "Array Buffer cannot be Shared", which is why I added this.  Which seems like a deliberate choice rather than a bug, but it may be something the Chrome devs only intended to be a temporary restriction.

juj commented on src/preamble.js:
```diff
@@ -433,7 +433,15 @@ function UTF8ArrayToString(u8Array, idx) {
 
   if (endPtr - idx > 16 && u8Array.subarray && UTF8Decoder) {
-    return UTF8Decoder.decode(u8Array.subarray(idx, endPtr));
+    var buf;
```
(this line could be merged to below for `var buf = ...` in both cases to save a few chars, but not a biggie)

juj commented on src/preamble.js:
```diff
@@ -433,7 +433,15 @@ function UTF8ArrayToString(u8Array, idx) {
+#else
+    buf = u8Array.subarray(idx, endPtr);
+#endif
+    return UTF8Decoder.decode(buf);
```
Or even duplicate the `UTF8Decoder.decode()` in both branches to keep the generated chars to minimum at the expense of a tiny bit of (temporary) duplication

eclecticdave replied:
Fair point. I would expect Closure would eliminate that anyway, but it's still worth doing.

juj commented on src/shell.html:
```diff
@@ -1224,6 +1224,37 @@
         preRun: [],
         postRun: [],
+        fdBuf: ['', '', ''],
+        printChars: function(fd, ptr, len) {
```
Here I suppose we could do the `printChars: function(fd, buffer, ptr, len)` change mentioned to earlier to avoid having to generate garbage? And then inside the body do `buffer = buffer || HEAPU8;` so that a null coming from asm.js/wasm will refer to `HEAPU8`?

eclecticdave replied:
Modified printChars signature as you suggest, only with the parameter order changed so that ptr (renamed to str_or_offset) is the only mandatory parameter.

juj commented on src/shell.html:
```diff
@@ -1249,6 +1280,7 @@
             console.error(text);
           }
         },
+        */
```
This can be then deleted for good.

juj commented:
Thanks, I think this is a good direction, moving responsibility on the buffering out from the MEMFS filesystem into the shell, and this will be nicely compatible with ASMFS as well.

eclecticdave commented:
So, it looks like three of the failing tests are due to them redefining Module['print'] and expecting printf to be directed to the replacement function.  Which obviously no longer happens here.
This, I think, represents a breaking change for users who define custom print functions - they would need to either redirect printChars instead or else undefine printChars which results in the code falling back to calling 'print'.

Do we need an option in settings.py here, or is it enough to just document the new behaviour?

eclecticdave commented:
Hi - I think I've now managed to address most of the review comments and (finally) got all the tests to pass.  Please can you take another look and let me know what you think?
Cheers.

eclecticdave commented:
Hi, I realise everyone's really busy at the moment, but if I can just drop a quick 'ping' here so this doesn't get forgotten about?
Cheers, David.

kripken commented:
Sorry for the delay here. The large issue is that I don't think @juj will have time to help out with this, and he's really the only one that knows the pthreads stuff well, which I think is in the background here - how important is pthreads here?

juj commented on src/library_syscall.js:
```diff
@@ -808,7 +808,8 @@ var SyscallsLibrary = {
     var ret = 0;
-    if (!___syscall146.buffers) {
+#if USE_PRINTCHARS_FOR_STDOUT == 0    
+    if (!___syscall146.buffer) {
```
This looks like has dropped `s` at the end of the if? `buffers` instead of `buffer`?

juj replied:
Btw, does `#if !USE_PRINTCHARS_FOR_STDOUT` work?

juj commented on src/library_syscall.js:
```diff
@@ -821,12 +822,17 @@ var SyscallsLibrary = {
     for (var i = 0; i < iovcnt; i++) {
       var ptr = {{{ makeGetValue('iov', 'i*8', 'i32') }}};
       var len = {{{ makeGetValue('iov', 'i*8 + 4', 'i32') }}};
+#if USE_PRINTCHARS_FOR_STDOUT == 0  
```
This might be nicer to read in positive direction, i.e. `#if USE_PRINTCHARS_FOR_STDOUT` instead of the negative?

juj commented on src/library_tty.js:
```diff
@@ -63,13 +63,22 @@ mergeInto(LibraryManager.library, {
         return bytesRead;
       },
       write: function(stream, buffer, offset, length, pos) {
+#if USE_PRINTCHARS_FOR_STDOUT == 1
```
In general we have had an idiom to just do `#if USE_PRINTCHARS_FOR_STDOUT`  rather than `#if USE_PRINTCHARS_FOR_STDOUT == 1` in scenarios where the intent is not to explicitly stress the value of `1`.

juj commented on src/library_tty.js:
```diff
@@ -153,8 +163,13 @@ mergeInto(LibraryManager.library, {
           tty.output = [];
-        } else {
-          if (val != 0) tty.output.push(val); // val == 0 would cut text output off in the middle.
+        } else if (val != 0) {
```
This change looks like intent to just micro-optimize away characters, in that case, I think we can even do `} else if (val) {` and drop the `val != 0`, although either way is just fine.

juj commented on src/library_tty.js:
```diff
@@ -153,8 +163,13 @@ mergeInto(LibraryManager.library, {
+        }
+      },
+      write: function(offset, length, buffer) {
+        if (Module['printChars'] != undefined && length) {
```
Can do ` if (Module['printChars']` instead of ` if (Module['printChars'] != undefined`

juj commented on src/library_tty.js:
```diff
@@ -169,8 +184,13 @@ mergeInto(LibraryManager.library, {
           tty.output = [];
-        } else {
-          if (val != 0) tty.output.push(val);
+        } else if (val != 0) {
```
Here can also `else if (val)`

juj commented on src/preamble.js:
```diff
@@ -419,21 +419,31 @@ function stringToAscii(str, outPtr) {
 var UTF8Decoder = typeof TextDecoder !== 'undefined' ? new TextDecoder('utf8') : undefined;
 #endif
-function UTF8ArrayToString(u8Array, idx) {
+function UTF8ArrayToString(u8Array, idx, len) {
```
This kind of change to add a `len` has been requested a number of times, and there is a very important subtlety to consider to well-define the `len` argument: what is the semantic that should occur when `len` is specified, but the input UTF8 array contains a null byte in the middle before `len` bytes are reached? Should a) the decoding continue past that null byte (writing the [null character](https://developer.mozilla.org/ro/docs/Web/JavaScript/Reference/Global_Objects/String) in the JS string) in the middle, up to exactly `len` bytes? or should b) the decoding stop at the null byte, and discard the remaining bytes (treating `len` rather more as a `maxLen` parameter).

Currently the changes in this PR implement b), but there is a problem that `TextDecoder` API actually implements a), so the implementation in this form has a bug that strings longer than 16 chars would get the behavior a), but strings shorter than that would get behavior b). I.e. TextDecoder is defined to allow null characters inside while converting, check this example:

```js
var d = new TextDecoder('utf-8');
var b = new Uint8Array([0x61, 0x6c, 0, 0x62]);
console.log(d.decode(b));
```

prints `al�b` (`al<nul>b`) in the browser.

If we want to define this as b), then we should make sure never to route to TextDecoder when `len` is specified, and perhaps rename `len` to `maxLen`. I'm leaning towards the semantic b) since that is more useful from security perspective, i.e. one can pass `UTF8ArrayToString` to convert strings with a predetermined known maximum buffer length, without having to count where the exact position of zero will be.

eclecticdave replied:
Fair enough - I've implemented this by modifying the while loop to always search until a nul character is found or maxLen is reached.  This + TextDecoder should be quicker that doing it ourselves.

juj commented on src/preamble.js:
```diff
@@ -419,21 +419,31 @@ function stringToAscii(str, outPtr) {
+  else endPtr += len;
 
   if (endPtr - idx > 16 && u8Array.subarray && UTF8Decoder) {
+#ifdef USE_PTHREADS
```
`#if USE_PTHREADS` instead of `#ifdef USE_PTHREADS`

eclecticdave replied:
np, though I think the preprocessor doesn't care - it looks like it would accept any token starting with 'if' in fact :-)

juj commented on src/settings.js:
```diff
@@ -877,3 +877,8 @@ var MEM_INIT_IN_WASM = 0; // for internal use only
 var SUPPORT_BASE64_EMBEDDING = 0; // If set to 1, src/base64Utils.js will be included in the bundle.
                                   // This is set internally when needed (SINGLE_FILE)
+
+var USE_PRINTCHARS_FOR_STDOUT = 1; // If set to 0, printf, puts and other stdio output functions will
```
Oh hmm, the original motivation was to be able to avoid having to have a preprocessor defines like this at all.

However it seems that it is too difficult to avoid. Perhaps we should rename this to `var UNBUFFERED_STDIO = 1;` instead to convey what the intent between these two sets of functions is.

Although, I'm now thinking if we are in a scenario where we do want to use a `-s` option for this, then perhaps we might be able to do a full breaking change by overloading `Module.printChars` to just be `Module.print`, which reads much nicer(?)

eclecticdave replied:
Hmm, my slight concern with UNBUFFERED_STDIO is it is still buffering in musl, so you have to call fflush() to get a string printed without a line-terminator.  Would this be confusing?

Not that I have any better suggestions - perhaps PRINT_PARTIAL_LINES or reverse it and have PRINT_FULL_LINES_ONLY=0?

+1 for overriding Module.print though.

eclecticdave replied:
On reflection your alternative suggestion of UNBUFFERED_PRINT will probably be fine - the comment describing it in settings.js can make it clear that it doesn't turn off libc stdio buffering.

juj commented on src/shell.html:
```diff
@@ -1224,30 +1224,36 @@
-          } else {
-            console.error(text);
+        fdBuf: ['', '', ''],
+        printChars: function(str_or_offset, fd, len, buffer) {
```
Convention is to use camel case in JS code: `strOrOffset`

juj commented on tests/cstdio/test_remove.cpp:
```diff
@@ -52,7 +52,7 @@ void test() {
   assert(!err);
 
-  std::cout << "success";
+  std::cout << "success" << std::endl;
```
Hmm, what made this test change necessary? Since we are making stdio more unbuffered than before, this feels like if something, this change would be going to the opposite direction?

eclecticdave replied:
Previously all writes to std::cout were adding a newline automatically - I was sure there was a bug for that, but I can't seem to find it now?  Anyway the changes in this PR fix that so that you have to explicitly output a newline.

I could have changed the test output, but decided to change to code in this case - I can do the former instead if that's clearer?

juj replied:
Wow, that sounds like a tricky bug indeed. Good to get that one fixed. Changing the test or the test output is fine either way, so this looks good.

juj commented on tests/fs/test_trackingdelegate.out:
```diff
@@ -4,8 +4,8 @@ Wrote to file "/file.txt"
-Wrote to file "/dev/tty"
-File read returned 'hello!'
+File read returned 'hello!Wrote to file "/dev/tty"
+'
```
What happened here?

eclecticdave replied:
Yes, this one's a bit kooky.  Internally 'printf', for reasons known only to itself, splits its calls to 'write' into two, in this case for the first 26 characters followed by the remaining 2 chars.  So the delegate function is called after the first 26 characters have been written.  The original code buffers this though, so output from the delegate function is printed first, followed by the printf output.  With the changes it prints the first 26 characters of the printf string, followed by the delegate function output then the last two characters.

In practice I don't think this behaviour change is necessarily a problem, since in a real program, if the delegate functions are used, I'd imagine they probably wouldn't print anything.

juj replied:
Oh gotcha, this looks fine then. Good sleuthing!

juj commented on tests/fs_after_main.cpp:
```diff
@@ -38,29 +38,51 @@ int main() {
   fclose(f);
   printf("Looping...\n");
   EM_ASM({
+#if USE_PRINTCHARS_FOR_STDOUT == 0
```
These would be nice to also read `#if USE_PRINTCHARS_FOR_STDOUT` to read in the positive direction.

juj commented on tests/module/test_stdin2.js:
```diff
@@ -0,0 +1,18179 @@
+// The Module object: Our interface to the outside world. We import
```
Adding the file `tests/module/test_stdin2.js` into the PR looks accidental, this file seems to be a compiled output code from Emscripten, rather than a source file? Can you rewrite the commit history to remove this file?

eclecticdave replied:
I've deleted it - I'm a little bit nervous rewriting the history since I read somewhere the Github can lose comments it you change the commit guids.

If you do a 'squeeze' when this is merged though that should take of it, I think.

juj commented on tests/test_core.py:
```diff
@@ -1162,9 +1162,9 @@ def test_exceptions_uncaught_2(self):
+            std::cout << "ERROR: uncaught_exception still set." << std::endl;
           else
-            std::cout << "OK";
+            std::cout << "OK" << std::endl;
```
The addition of `std::endl`s here looks also odd to me? Was there something that did not work here without these?

eclecticdave replied:
See above

juj commented:
> Do we need an option in settings.py here, or is it enough to just document the new behaviour?

Given that a new setting was added, I am thinking that perhaps we should then just break the semantic of the existing `Module.print` function and gate that breakage on the value of that settings.py option? That is, the semantic change is that if `-s UNBUFFERED_STDIO=1` (or `-s UNBUFFERED_PRINT=1` a better name?), then one needs to call `Module.print('hello\n');` rather than `Module.print('hello');` to print a line? Or we could also keep using the `printChars` name, though having two names might not be necessary since only one of them will ever be used?

juj commented:
> how important is pthreads here?

@kripken : for this specific PR, there's nothing that is pthreads related, except for that one line that papers over the fact that one can't pass SharedArrayBuffers into TextDecoder API in Chrome, only regular ArrayBuffers work, so it clones a temp buffer from SAB -> AB when needed.

Apart from that line, this is all regular singlethreaded work, with the only idea being that instead of having `src/library_fs.js` own the buffering up of full lines of text before flushing, we would instead move that responsibility to `Module.print()` function instead. This is nice also for singlethreaded builds, since for example printing data to a text box DOM element on a web page, or to Node.js `process.stdout.write()`,  or to Firefox debug `window.dump()` does not require line buffering, whereas `console.log()` does; so making it the responsibility of the `Module.print()` function to choose how whether to line buffer or not allows skipping that when printing to an output that does not need it.

eclecticdave commented:
OK, I've fixed some of the smaller things, but it looks to me that to overload the Module.print function based on the UNBUFFERED_PRINT setting, as discussed in @juj's latest review, is going to need a preprocessor to handle #if/#else on shell.html.

However as it happens, I decided this would be useful anyway for when I resubmitted my keyboard handler changes, so I already had a stab at it.  It's not entirely finished yet, but I'll create a PR for it for discussion purposes.

juj commented on src/preamble.js:
```diff
@@ -419,21 +419,33 @@ function stringToAscii(str, outPtr) {
+  while (u8Array[endPtr]) {
+    if (maxLen != undefined && (idx + maxLen) == endPtr) break;
+    ++endPtr;
+  }
```
This can be a very performance sensitive spot, so let's instead make this read

```javascript
if (maxLen != undefined) {
  var maxPtr = idx + maxLen;
  while (endPtr < maxPtr && u8Array[endPtr]) ++endPtr;
} else {
  while(u8Array[endPtr]) ++endPtr;
}
```

so that the heavy inner loop that has to scan through the whole string length stays as slim as possible?

juj commented on src/shell.html:
```diff
@@ -1224,30 +1224,36 @@
-          } else {
-            console.error(text);
+        fdBuf: ['', '', ''],
+        printChars: function(strOrOffset, fd, len, buffer) {
```
The input parameters and their optionality feels still a bit off. Usually in APIs when there is a (offset, length) pair being handled, the offset and length are usually two consecutive parameters, but here we have a `offset, <unrelated>, length` syntax. I think that we should define this as having the following possible usage forms:

```javascript
// String-based overload:
Module.printChars(1, "Hello stdout!");
Module.printChars(2, "Hello stderr!");
// ArrayBuffer-based overload:
Module.printChars(1, 10/*offset*/, 63/*length*/, HEAPU8); // Prints UTF8 string in HEAPU8[10]...HEAPU8[10+63]
Module.printChars(1, 10, 63); // When source buffer is not specified, then HEAPU8 is used.
Module.printChars(1, 10); // length is also optional, by virtue of how UTF8ArrayToString() is being changed in this PR.
Module.printChars(1, 10, undefined, myArray); // length can be passed as undefined in the middle to specify unbounded read from a custom array.
```

Then the code would read

```javascript
printChars: function(fd, str, len, buffer) {
  if (typeof(str) === 'number') str = UTF8ArrayToString(buffer || HEAPU8, str, len);
  ...
```

and `fd` would never be optional.

juj commented on src/shell.js:
```diff
@@ -273,6 +273,20 @@ else {
+    putstr(str);
+  }
+}
+
```
shell.js already supports preprocessor, so this whole thing can be wrapped inside a `#if USE_PRINTCHARS_FOR_STDOUT`

eclecticdave commented:
Hi,
I've think I've implemented these latest suggestions - and I've renamed the setting I've added to UNBUFFERED_PRINT.

To clarify the current position:

Module.print (and printErr)  operate in 'buffered' mode, just like always, only printing when a newline character is reached.
Module.printChars operates in 'unbuffered' mode i.e. it prints strings straightaway without waiting for a newline.
Setting UNBUFFERED_PRINT controls whether printf, puts etc. call 'print' or 'printChars'.

The one outstanding question (as far as I'm aware) is whether to change things so 'print' itself switches between 'buffered' and 'unbuffered' mode.  As discussed above to do this for the browser version we need a preprocessor for html files - so I've submitted PR 6510 for comments.

eclecticdave commented:
Looks like I was stung by #6450 on the CI - I'll merge incoming and see if that helps.

eclecticdave commented:
The browser tests have failed with 'no http activity' - which sounds like an issue with the test environment?

eclecticdave commented:
So, this has required a little bit of a rejig as a result of the introduction of the 'out' and 'err' functions... since the merge history was complicated enough as it was, I've opted to do a rebase which should hopefully simplify merging this PR.

In addition I've opted to simplify the code a little - I haven't included a combined single function to replace Module.print and printErr in this latest iteration.  I still like the idea of such a function, but it no longer seemed to 'fit' quite as well alongside the new 'out' and 'err' functions so I've left it out for the time being at least.

cancerberoSgx commented:
any news on this ? Thanks!

eclecticdave commented:
Hi @cancerberoSgx,

Apologies - I've had some distractions in recent months.  The code in this PR should be generally working - the main reason for the large number of test failures (particularly the browser tests) seem to be where the tests redirect the 'out' and 'err' functions expecting 'printf' and the like to call the replacements - which no longer happens with this PR.  I need to find time to sit down and work my way through fixing these.
